### ISSUE TITLE
FreeBSD port

### ISSUE THREAD

**Issue by idefix6** [gogits/gogs#952]
Hi,

I just tried gogs and it look really nice and I like it. As I work on FreeBSD servers I would like to create a FreeBSD port for it to get it easily install able via ports system.
I saw that already a startup script for gogs is existing.
Is here already someone working on a port?
Is it ok if I start to create a port for it?

Thanks
Matthias

**Comment by phsmit** [comment]
I am also using gogs on FreeBSD and have also been looking how to make it a port.

One problem at this moment is still that a lot of files are dependent on the location of the gogs binary, meaning that is hard to find an appropriate place in the filesystem. This will change in the coming versions (I think @Unknwon mentioned he wants static files in the binary for version 0.6). This is the main reason I have not made a port yet, but if you are inspired, go ahead!!!

**Comment by Unknwon** [comment]
Thanks for asking!

> Is here already someone working on a port?

Bases on @phsmit , no one yet.

> Is it ok if I start to create a port for it?

It would be awesome!

Ping us if you have any question/issue, we are very happy to help as much as we can. :star:

**Comment by idefix6** [comment]
Thanks a lot for help. I'm completely new to go and the complete structure which is behind is not 100% clear to me. I started with the port: https://github.com/idefix6/gogs-freebsdport

What I do not understand is, that all the modules are part of gogs, like go-sql-driver and many many more. Is it not more clean to have these modules as separate package installed on the standard path like:
/usr/local/share/go/pkg/freebsd_amd64/github.com/go-sql-driver/mysql.a
/usr/local/share/go/src/github.com/go-sql-driver/mysql/*
(this seems it least to be the standard on FreeBSD, but as I said, I have no experience with go, only with normal Makefile and cmake projects).
And gogs will use this module on this path?
The binary of gogs will then be installed in /usr/local/bin or in any were go expect it (there is is a binary and I think the best place is anywhere in the file system were the normal PATH points to).

**Comment by phsmit** [comment]
Some hints:
 - Compiling in go will give you a static executable. There is nothing you can do about it, and it is a good thing. This means you will not need any runtime dependencies at all.
- At compile time you can use other static compiled libraries, but this is not necessary (or go will do that under the hood). I'm not sure what method is used atm by other go-ports on freebsd, but most likely you should just fetch them all yourself.
- The install location of the binary should ideally be in /usr/local/bin , but take in account that at this moment at least the "canonical" conf file must be located relative to the executable (/urs/local/bin/conf/app.ini). This should change in a next version. You can already put the real config in /usr/local/etc and the static stuff in /usr/local/share

**Comment by lambdaupb** [comment]
I have gogs running on FreeBSD, it just works IIRC.

here is my 5minute hacked together init (rc) file:

```
#!/bin/sh
# PROVIDE: gogs
# REQUIRE: DAEMON NETWORKING postgres
# BEFORE:
# KEYWORD:

. /etc/rc.subr

name=gogs
rcvar=${name}_enable

load_rc_config $name

user="gogs"
user_home="/usr/home/gogs"
command="./gogs web"
pidfile="/var/run/${name}.pid"

start_cmd="${name}_start"
stop_cmd="${name}_stop"
status_cmd="${name}_status"

gogs_start()
{
        export USER=$user
        export HOME=$user_home
        cd $HOME/gogs
        if checkyesno ${rcvar}; then
                /usr/sbin/daemon -u ${user} -p ${pidfile} ${command} > /var/log/gogs.log 2>&1
        fi
}

gogs_stop()
{
        if [ -f ${pidfile} ]; then
                kill `cat ${pidfile}`
        fi
}

gogs_status()
{
        if [ ! -f ${pidfile} ] || kill -0 `cat ${pidfile}`; then
                echo "It appears ${name} is not running."
        else
                echo "${name} seems to be running."
        fi
}

run_rc_command $1
```

**Comment by Unknwon** [comment]
@lambdaupb  thanks for the script!

**Comment by sashkab** [comment]
I've slightly modified init file to follow same logic, as install guide on http://gogs.io/docs/installation/install_from_source.html. Here is my file:

```sh
#!/bin/sh
# PROVIDE: gogs
# REQUIRE: DAEMON NETWORKING postgres nginx
# BEFORE:
# KEYWORD:

. /etc/rc.subr

name=gogs
rcvar=${name}_enable

load_rc_config $name

user="git"
user_home="/home/git"
command="./gogs web"
pidfile="/var/run/${name}.pid"

start_cmd="${name}_start"
stop_cmd="${name}_stop"
status_cmd="${name}_status"

gogs_start()
{
        export USER=$user
        export HOME=$user_home
        export GOPATH=$HOME/go
        cd $GOPATH/src/github.com/gogits/gogs
        if checkyesno ${rcvar}; then
                /usr/sbin/daemon -u ${user} -p ${pidfile} ${command} > /var/log/gogs.log 2>&1
        fi
}

gogs_stop()
{
        if [ -f ${pidfile} ]; then
                kill `cat ${pidfile}`
        fi
}

gogs_status()
{
        if [ ! -f ${pidfile} ] || kill -0 `cat ${pidfile}`; then
                echo "It appears ${name} is not running."
        else
                echo "${name} seems to be running."
        fi
}

run_rc_command $1
```

**Comment by idefix6** [comment]
Thanks for the script, I added it to the maybe new port for FreeBSD, I hope that is ok for you?
https://github.com/idefix6/gogs-freebsdport/commit/4d822619ebef2ee541204339783b0e2f88f4fc7b

**Comment by sashkab** [comment]
@idefix not a problem at all, I wasn't original author to start with :)

**Comment by lambdaupb** [comment]
@idefix6 Yes, consider it BSD licensed, (specifically the one used for FreeBSDs base init scripts)

**Comment by tobiastom** [comment]
A FreeBSD port would be awesome.

**Comment by nd2s** [comment]
+1

**Comment by Unknwon** [comment]
Any updates?

**Comment by idefix6** [comment]
I continued with the port, but I have problems compiling it. As I was never in touch with go before, I would need some help here.
I updated the port to use gogs 0.7.33.
But if I try to do a `go get -x` I get:
```
WORK=/tmp/go-build982760472
mkdir -p $WORK/github.com/gogits/gogs/models/_obj/
mkdir -p $WORK/github.com/gogits/gogs/
cd /wrkdirs/usr/ports/devel/gogs/work/src/github.com/gogits/gogs/models
/usr/local/go/pkg/tool/freebsd_amd64/compile -o $WORK/github.com/gogits/gogs/models.a -trimpath $WORK -p github.com/gogits/gogs/models -complete -buildid 90de7610b6891247c052128c457749c586871a96 -D _/wrkdirs/usr/ports/devel/gogs/work/src/github.com/gogits/gogs/models -I $WORK -I /wrkdirs/usr/ports/devel/gogs/work/pkg/freebsd_amd64 -pack ./access.go ./action.go ./admin.go ./error.go ./git_diff.go ./issue.go ./login.go ./models.go ./org.go ./publickey.go ./pull.go ./release.go ./repo.go ./token.go ./update.go ./user.go ./webhook.go ./webhook_slack.go ./wiki.go
# github.com/gogits/gogs/models
models/repo.go:101: assignment count mismatch: 2 = 1
models/repo.go:344: not enough arguments in call to "github.com/gogits/git-shell".Clone
```
Had anyone an idea, how to get that fix gogs can compile?

Thanks.

**Comment by thehowl** [comment]
What is `-x` for? I don't see it in the options for `go get`. You might as well have meant to use `go get -u` (makes sure all dependencies are up-to-date)

**Comment by Unknwon** [comment]
Thanks @thehowl , he made the point :smile:

**Comment by idefix6** [comment]
I just saw that already a port is ready and can be tested:
https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=205283

**Comment by Unknwon** [comment]
@idefix6 thanks!

**Comment by idefix6** [comment]
A first test package is now available: https://pkg.fechner.net/packages/102amd64-default/All/gogs-0.151216.txz

**Comment by Unknwon** [comment]
@idefix6 Does it mean Gogs now can run on FreeBSD as other OSs?

**Comment by idefix6** [comment]
The port is ready and works, but is on in the FreeBSD repository yet.
If it is in the FreeBSD repository a simple ```pkg install gogs``` is enough.

**Comment by Unknwon** [comment]
Thanks!

> The port is ready and works, but is not in the FreeBSD repository yet.

How to achieve that?

**Comment by idefix6** [comment]
we have to wait till all patches are accepted. In the meantime you can use the package I created to get it easily installed.

**Comment by Unknwon** [comment]
@idefix6 OK... I don't use FreeBSD so I can't test...

**Comment by Asara** [comment]
As an update for anyone who was interested, gogs still isn't in the fBSD repo, but can be installed easily using: https://github.com/gogits/gogs/issues/952#issuecomment-84351879

**Comment by tobiastom** [comment]
@idefix6 is there a pull request to follow your progress?

**Comment by idefix6** [comment]
There is already a package ready that can be tested, see here: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=205283#c34

**Comment by optikfluffel** [comment]
Any updates on this?
